File: dotnet/Aoc/Aoc.Spaceship/Security/PasswordGenerator.cs
using System.Collections.Generic;

namespace Aoc.Spaceship.Security
{
    public class PasswordGenerator
    {
        public bool Check(int candidate)
        {
            var passesTests = NumberHasDuplicateDigit(candidate);
            passesTests = passesTests && DigitsAlwaysIncrease(candidate);
            return passesTests;
        }

        public bool NumberHasDuplicateDigit(int candidate)
        {
            var candidateString = candidate.ToString();
            var matchingDigits = new List<char>();
            var hasDouble = false;
            foreach (var candidateChar in candidateString.ToCharArray())
            {
                if (matchingDigits.Count > 0)
                {
                    if (matchingDigits[0] == candidateChar)
                    {
                        matchingDigits.Add(candidateChar);
                    }
                    else
                    {
                        if (matchingDigits.Count == 2)
                            hasDouble = true;
                        matchingDigits.Clear();
                        matchingDigits.Add(candidateChar);
                    }
                }
                else
                {
                    matchingDigits.Add(candidateChar);
                }
            }

            if (matchingDigits.Count == 2)
                hasDouble = true;
            return hasDouble;
        }

        public bool DigitsAlwaysIncrease(int candidate)
        {
            var candidateString = candidate.ToString();
            var biggestNumber = 0;
            foreach (var candidateChar in candidateString.ToCharArray())
            {
                var digit = int.Parse(candidateChar.ToString());
                if (biggestNumber > digit)
                    return false;
                biggestNumber = digit;
            }
            return true;
        }

        public List<int> GetCandidatePasswords(int startIndex, int endIndex)
        {
            var candidates = new List<int>();
            for (var candidate = startIndex; candidate <= endIndex; candidate++)
            {
                if (Check(candidate))
                    candidates.Add(candidate);
            }
            return candidates;
        }
    }
}

File: python/aoc/spaceship/security/password_generator.py
def validate_password(candidate: int) -> bool:
    return number_has_duplicate_digit(candidate) and digits_always_increase(candidate)


def get_candidate_passwords(starting_number: int, ending_number: int) -> list[int]:
    candidates: list[int] = []
    for candidate in range(starting_number, ending_number):
        if validate_password(candidate):
            candidates.append(candidate)
    return candidates


def number_has_duplicate_digit(candidate: int) -> bool:
    matching_digits = []
    has_double = False
    for candidate_char in list(str(candidate)):
        if (len(matching_digits) == 0) or (matching_digits[0] == candidate_char):
            matching_digits.append(candidate_char)
        else:
            if len(matching_digits) == 2:
                has_double = True
            matching_digits = [candidate_char]
    if len(matching_digits) == 2:
        has_double = True
    return has_double


def digits_always_increase(candidate: int) -> bool:
    candidate_string = str(candidate)
    biggest_number = 0
    for candidateChar in list(candidate_string):
        digit = int(candidateChar)
        if biggest_number > digit:
            return False
        biggest_number = digit
    return True

